UOP1_Project/Assets/Scripts/StateMachine/Editor/Templates/StateAction.txt

```diff
@@ -0,0 +1,28 @@
+using UnityEngine;
+using UOP1.StateMachine;
+using UOP1.StateMachine.ScriptableObjects;
+
+[CreateAssetMenu(fileName = "#RUNTIMENAME#", menuName = "State Machines/Actions/#RUNTIMENAME_WITH_SPACES#")]
+public class #SCRIPTNAME# : StateActionSO
+{
+	protected override StateAction CreateAction() => new #RUNTIMENAME#();
+}
+
+public class #RUNTIMENAME# : StateAction
+{
+	public override void Awake(StateMachine stateMachine)
+	{
+	}
+		
+	public override void OnUpdate()
+	{
+	}
+	
+	// public override void OnStateEnter()
+	// {
+	// }
+	
+	// public override void OnStateExit()
+	// {
+	// }
```

Comment:
I'd personally uncomment these lines.
_Commented_ results in:
- If you want to keep them, you need to uncomment them
- If you don't need them, you might as well remove them

_Uncommented_ results in that at least if you need them, they are ready to use. So overall they save everyone more time :)

Reply:
Makes sense, I guess I just wanted to show that you _could_ remove them, unlike the rest. Since we're at it, I always felt unsure whether my choice of abstract/virtual in these methods were ok, what do you think?

Reply:
> Since we're at it, I always felt unsure whether my choice of abstract/virtual in these methods were ok, what do you think?

They're good for classes, but if we're going to use structs soon, we would not be able to use virtual methods. Maybe we should make IStateEnter and IStateExit interfaces instead?

Reply:
I thought we weren't going with structs in the end. Maybe bring this over to the thread, but, what would be the benefits of using structs at this point?

Reply:
I think I've already described their benefits in the thread. I'll try to make a PR with them later next week, listing all the cons and pros with more in-depth examples

Reply:
Sorry @neon-age, but we need to move on from shaking the fundamentals of the StateMachine. Feel free to bring the argument in the thread, but I would advice not to jump into great refactorings...

Even with this "non-breaking" PR, I had to change 25 files, not to mention test. Plus, we need to write docs on the StateMachine so we need to kinda set it in stone.